Payment requests

andrewkozlik opened the pull request:
This draft PR implements payment requests in Trezor T. Below is a screenshot of the payment request confirmation screen. This replaces the output confirmation screen:

![image](https://user-images.githubusercontent.com/42678794/105690872-ced88d00-5efc-11eb-96fa-3e442931512d.png)

I still need to add some device tests and maybe some configuration options:

Do we want to allow users to disable this feature?
- Disabling it for CoinJoin makes no sense because it just checks the authenticity of the coordinator.
- Some users might want to disable it for standard transactions, because although they trust the firmware, they might not want SatoshiLabs to have control over what is displayed on the output confirmation screen.

I see two options:
1. Users could configure Trezor to disable this feature in standard transactions, i.e. not to replace the address with the recipient name on the output confirmation screen.
2. When confirming the payment request, the user could walk through the addresses of all the outputs that are hidden under the payment request. Implementation-wise this means that Trezor would need to maintain a list of output indices in RAM for each payment request being processed. Probably not a problem, we already do something similar for segwit inputs.

Do we want to add a feature flag in GetFeatures, so that Suite can tell whether the feature is available? If it won't be possible to disable this feature, then Suite can just decide based on the firmware version.

matejcik commented:
> Do we want to allow users to disable this feature?

a better solution IMHO is allowing showing the full address. ping @matejzak to take this into consideration?

matejcik commented:
...which is this, i presume?

> 2\. When confirming the payment request, the user could walk through the addresses of all the outputs that are hidden under the payment request. Implementation-wise this means that Trezor would need to maintain a list of output indices in RAM for each payment request being processed. Probably not a problem, we already do something similar for segwit inputs.

but I don't fully understand. This is for the case where multiple payment requests are intermingled?

tsusanka commented:
We have discussed this in Product and yes, let's go with (2). User will be able to see the address when requested. @matejzak will prepare design for this. I have also created an [issue](https://github.com/trezor/trezor-firmware/issue/1430) to track this better and included the design request (product spec) there. It should be done in a few days.

andrewkozlik commented:
> but I don't fully understand. This is for the case where multiple payment requests are intermingled?

Yes, I am working with the assumption that the outputs of the payment requests may be interleaved. I have done some rethinking of the proposed flow and implemented it differently in a set of fixup commits. The idea is to remove the `amount` and `hash_outputs` fields:
https://github.com/trezor/trezor-firmware/pull/1429/commits/53ed3970cb76e035170ae56cd8b5cce4fc512da9#diff-4568999ce3d84d878396d98f640d2cab8ee735bc568324e7e387ef4254c840a5R405-R409
- Advantage: Gets rid of `expected_hash_outputs` and `expected_amount`, which saves some memory and makes the logic in Trezor a bit simpler, which in turn implies a smaller surface for vulnerabilities.
- Advantage: Makes the API simpler for Suite, because less information needs to be provided.
- Disadvantage: All payment requests are requested and confirmed at the end of [`step2_approve_outputs`](https://github.com/trezor/trezor-firmware/pull/1429/files#diff-59c3bd122a15262a3c773033fce68de1d398eaf8f401cecc1b775bb9a36a2797R153-R157), which means that the `HashWriters` have to be kept in memory for the entire time. Plus if we want the user to be able to walk through the addresses as proposed above, then we also need to keep the set of output indices in memory.

The disadvantage above is why I initially proposed the design the way I did. The reason why I changed my mind is that if the number of payment requests is small, then this is not a problem, and the advantages far outweigh the disadvantage. Currently we have no use-case for a large number of payment requests. If the number of payment requests becomes large for some application, then this can be easily solved by adding a flag to the `TxOutput`, e.g. `bool last_in_payment_request`, saying "this is the last output belonging to the given payment request". When that flag is set, the payment request will be fetched and confirmed immediately after receiving that output. In fact we could implement this flag right now and make it part of the required flow, i.e. Suite would always have to mark the last output of a payment request.

matejcik reviewed (commented):
Looks reasonable overall. I'm not particularly happy about post-loading the payment requests though, nor about the `last_in_payment_request` thing.
Remind me, how important is it to allow interleaving of payment request outputs? It brings a lot of complications.
If I were not trying to shoehorn this into the old signing flow, I'd ask the host for a list of payment requests, and query the list of outputs for each in sequence. That way you'd keep the payment request in memory exactly as long as you need, and it would also make reviewing the output addresses easier.

One more comment: verifying locality of the LTC/Dash/Doge addresses is nice and all, but it only works for Bitcoin-like coins, which seems limiting.

andrewkozlik commented on common/protob/messages-bitcoin.proto:
```diff
@@ -396,6 +399,24 @@ message PrevOutput {
+    optional bytes nonce = 3;           // the nonce used in the signature computation
+    required bytes signature = 4;       // the recipient’s signature of the paymentRequestDigest
+
+    message Memo {
```
https://github.com/trezor/trezor-firmware/pull/1429/commits/a3a447fe60c69355a8444e86ab3c5da0fb2ee5ea

andrewkozlik commented on python/src/trezorlib/btc.py:
```diff
@@ -154,6 +154,11 @@ def get_ownership_proof(
 
 
+@expect(messages.Nonce, field="nonce")
+def get_nonce(client):
```
https://github.com/trezor/trezor-firmware/pull/1429/commits/a4572c0980194081b5f930ec30297cbe58d1213f

andrewkozlik commented on tests/device_tests/test_msg_signtx_payreq.py:
```diff
@@ -0,0 +1,277 @@
+        txo.payment_request = None
+
+    payment_reqs = []
+    for i, (txo_indices, hash_outputs, memos) in enumerate(payment_request_params):
```
https://github.com/trezor/trezor-firmware/pull/1429/commits/234f5c62a0cc954959914fe142ca44605381ffa4

andrewkozlik commented on core/src/apps/bitcoin/sign_tx/approvers.py:
```diff
@@ -58,10 +61,27 @@ def add_external_input(self, txi: TxInput) -> None:
+            txo.payment_request is not None
+            and txo.payment_request not in self.payment_requests
+        ):
+            self.payment_requests[txo.payment_request] = PaymentRequest()
```
Very good point. Done https://github.com/trezor/trezor-firmware/pull/1429/commits/4360111c79ccd88dbd34e8f18201b6993044903d. I could also move the code from payment_request.py into verification.py. On the other hand the payment request stuff is only used in sign_tx while verification.py is one directory-level up.

matejcik replied:
that sounds more like a name clash than a good structure? it makes sense to keep payment request related code together, even if it's called "verifier". Maybe the "verification" in verification.py should be less generic instead?
(but I don't think there's actual need for that, not in this PR in any case)

andrewkozlik replied:
Agreed. Let's leave as is.

andrewkozlik commented on core/src/apps/bitcoin/sign_tx/bitcoin.py:
```diff
@@ -150,6 +150,11 @@ async def step2_approve_outputs(self) -> None:
             await self.approve_output(txo, script_pubkey, orig_txo)
 
+        # Approve payment requests.
+        for index in self.approver.payment_requests:
```
> Remind me, how important is it to allow interleaving of payment request outputs? It brings a lot of complications.

The only application that I had in mind was WabiSabi. The way I understand it, this would be a CoinJoin-like transaction which would be used to pay to another party. In order to reduce linkability of the payment input and payment output, the recipient might need to create multiple outputs, all of which need to be unlinkable to an outside observer (and even to the coordinator), so they shouldn't be clumped one after another, but randomly mixed among the others. For the user who is signing the transaction it would be tricky to verify the outputs manually and it would be much nicer to hide them under a payment request. But we should also consider that since the whole scheme is CoinJoin-like, the payment requests will probably need to be preauthorized, now that I am thinking about it.

So this capability seems to be quite far at the moment. I don't have anything against simplifying the current implementation as long as the protocol allows us to add this in the future in a backwards-compatible way.

> One more comment: verifying locality of the LTC/Dash/Doge addresses is nice and all, but it only works for Bitcoin-like coins, which seems limiting.

Yeah, I am not sure how to deal with that elegantly. BTW, the payment request memos contain a scriptPubKey, which seemed like a more generic solution to me than an address. Any thoughts on that?

andrewkozlik replied:
Let's first discuss the overall flow before we get to this. Regarding point (a) I think it's more error-prone to add another variable. It violates the single-point-of-truth principle.

On a side note, because of the reviewing of output addresses, I am a bit worried we might need to implement (c). Another alternative would be to return the amount field to `TxAckPaymentRequest` and allow only one pass.

matejcik replied:
> Another alternative would be to return the amount field to `TxAckPaymentRequest` and allow only one pass.

how would that help? the current code is also one-pass

matejcik replied:
...ah, right, but then we can pre-load the payment request instead of post-load

andrewkozlik replied:
> how would that help? the current code is also one-pass

I was referring to the reviewing of output addresses. You can either preload the payment request and then allow the user to review the individual outputs or you can postload it and perhaps even allow the user to cycle through them repeatedly (like the XPUBs in GetAddress). The second option is a littlebit tricky implementation-wise, because you either have to keep them in memory or stream all of them to recompute the `hash_outputs` to make sure an attacker is not giving you a fake one.

andrewkozlik commented:
> Remind me, how important is it to allow interleaving of payment request outputs? It brings a lot of complications.

The only application that I had in mind was WabiSabi. The way I understand it, this would be a CoinJoin-like transaction which would be used to pay to another party. In order to reduce linkability of the payment input and payment output, the recipient might need to create multiple outputs, all of which need to be unlinkable to an outside observer (and even to the coordinator), so they shouldn't be clumped one after another, but randomly mixed among the others. For the user who is signing the transaction it would be tricky to verify the outputs manually and it would be much nicer to hide them under a payment request. But we should also consider that since the whole scheme is CoinJoin-like, the payment requests will probably need to be preauthorized, now that I am thinking about it.

So this capability seems to be quite far at the moment. I don't have anything against simplifying the current implementation as long as the protocol allows us to add this in the future in a backwards-compatible way.

> One more comment: verifying locality of the LTC/Dash/Doge addresses is nice and all, but it only works for Bitcoin-like coins, which seems limiting.

Yeah, I am not sure how to deal with that elegantly. BTW, the payment request memos contain a scriptPubKey, which seemed like a more generic solution to me than an address. Any thoughts on that?

andrewkozlik commented on core/src/apps/management/get_nonce.py:
```diff
@@ -0,0 +1,19 @@
+
+async def get_nonce(ctx: wire.Context, msg: GetNonce) -> Nonce:
+    nonce = random.bytes(32)
+    nonces: Set[bytes] = cache.get(cache.APP_COMMON_NONCES)
```
Very good point. Done in https://github.com/trezor/trezor-firmware/pull/1429/commits/585892a254d8b1f47b6742fc5a68bea469514f23, https://github.com/trezor/trezor-firmware/pull/1429/commits/ac1cff979ef35de95309463bac468141f26bc252.

andrewkozlik commented:
> > One more comment: verifying locality of the LTC/Dash/Doge addresses is nice and all, but it only works for Bitcoin-like coins, which seems limiting.
> 
> Yeah, I am not sure how to deal with that elegantly. BTW, the payment request memos contain a scriptPubKey, which seemed like a more generic solution to me than an address. Any thoughts on that?

So I am contemplating two possible solutions:
1. Change `PaymentRequestMemo.data` to be a public key instead of a scriptPubKey.
    - Remove `script_type` field from `PaymentRequestMemo`.
    - The Bitcoin app would still need to call some coin-dependent path validation, perhaps introducing unwanted coupling.
    - Not sure whether this makes sense for all cryptocurrencies.
2. Change `PaymentRequestMemo.data` to be an address instead of a scriptPubKey.
    - Remove `script_type` and `address_n` fields from `PaymentRequestMemo`.
    - Add a `mac` field to `PaymentRequestMemo`.
    - Add a `mac` field to all the `*Address` messages, so `*GetAddress` would return an `HMAC(key, slip44_coin_type || address)`, where `key` is a symmetric key derived from the seed. The host application would need to provide this MAC in the `PaymentRequestMemo` to prove that the address was obtained from Trezor.

The second solution seems like the most appealing to me at the moment.

andrewkozlik commented:
In https://github.com/trezor/trezor-firmware/pull/1429/commits/7f1a5ecdb3a88768e6d7cf52a928ffa97ebf1246 I added the `mac` field to the `Address` response message. But I think the change to the return type of `btc.get_address()` trezorlib is a problem. So I assume the best way to do this is to instead implement something like `btc.get_authenticated_address()`.

In https://github.com/trezor/trezor-firmware/pull/1429/commits/b4287fb6084eab735506e9dbda6b84c5840fab7e I reworked the payment request handling:
- All outputs belonging to a payment request have to be consecutive. I don't think this made the code any more legible though.
- The MAC is used to identify ownership of an address.
- The coin memo's don't just get silently checked now, but they show up on the screen as "Buying ...", so this information is no longer provided in the text memo. As @matejcik pointed out, this is better for translations. I use a specific memo type COIN_PURCHASE for this. So in the future if we have a use-case for silent verification, then we can add another type.
- Reorganized the way the signed data is hashed (hash_outputs come at the end now).

Would it make sense to have different protobuf messages for different types of memos? I am not sure whether protobuf is able to handle polymorphism.

matejcik commented:
> But I think the change to the return type of `btc.get_address()` trezorlib is a problem. So I assume the best way to do this is to instead implement something like `btc.get_authenticated_address()`.

this is reasonable

> Would it make sense to have different protobuf messages for different types of memos? I am not sure whether protobuf is able to handle polymorphism.

we do polymorphism by specifying multiple optional submessages:
```protobuf
message PaymentRequest {
	optional PaymentRequestPurchase purchase = 1;
	optional PaymentRequestCommitment commitment = 2;
	// ...
}
```
As to whether it makes sense, i don't really think so -- what would be the differences?
If the difference is which optional (sub)fields are required, then this doesn't buy all that much.

OTOH we could stream at the level of individual memos, where we can use a similar approach as `TxAckInput`, `TxAckOutput` etc.

Which also makes me think that given "hash_outputs" are at end, we can save memory by discarding the memos and only keeping the running hash, right? Which should alleviate some concerns re long-lived payment requests

andrewkozlik commented:
> we do polymorphism by specifying multiple optional submessages [...]
> If the difference is which optional (sub)fields are required, then this doesn't buy all that much.

As I thought. I agree it's not worth it. I can't think of a use-case, where there would be so many memos or where the memos would be so large that we would need to stream them.

> Which also makes me think that given "hash_outputs" are at end, we can save memory by discarding the memos and only keeping the running hash, right?

Yes, that is the reason I placed the field at the end.

andrewkozlik commented:
https://github.com/trezor/trezor-firmware/pull/1429/commits/6d75a78d56d699c1e9d0365e04e3f65d4c12899a: Adds `btc.get_authenticated_address()`.
https://github.com/trezor/trezor-firmware/pull/1429/commits/079bb97442518b884ee76c9cde1cbf6aed0edca2: Adds details button to payment request confirmation screen. Examples (no memo, UTF8_TEXT, COIN_PURCHASE):
![image](https://user-images.githubusercontent.com/42678794/107238102-bea9cd00-6a27-11eb-8baf-6fe1464b071d.png) ![image](https://user-images.githubusercontent.com/42678794/107238156-cf5a4300-6a27-11eb-977c-9c55ee4a2f95.png) ![image](https://user-images.githubusercontent.com/42678794/107238192-d97c4180-6a27-11eb-9910-38e5d3ba9da6.png)

I will add some more tests, squash the fixups and create a fresh PR.

andrewkozlik commented:
Replaced by https://github.com/trezor/trezor-firmware/pull/2069.
